[Debug info] Retain all type aliases in asttypes mode

adrian-prantl opened the pull request:
Bound generic types don't reference their type parameters in ASTTypes mode, so we need to artificially keep typealiases alive, since they can appear in reflection metadata.

rdar://143595521

adrian-prantl committed:
[Debug info] Emit bound generic class type parameters when emitting AST types

This ensures that debug info for these types is emitted even if this is the only
way they are references in the input program.

rdar://143595521

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

augusto2112 commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2182,8 +2182,14 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
+      // mode, so we need to artificially keep typealiases alive, since they can
+      // appear in reflection metadata.
+      if (Opts.DebugInfoLevel < IRGenDebugInfoLevel::DwarfTypes)
+        DBuilder.retainType(TypeDef);
```
I thought `retainType` would retain the type in the IR but not in DWARF, am I wrong?

adrian-prantl replied:
It does both:
https://github.com/llvm/llvm-project/blob/5d6d982df61d16b6d498e6d59dd91c059679d3d8/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#L1221

augusto2112 replied:
Hmm, does dsymutil keep them too? I think maybe that's what I was remembering. LGTM if yes

adrian-prantl replied:
It should not, that's a great point.

adrian-prantl replied:
Here is a better solution that builds on top of https://github.com/swiftlang/swift/pull/78965.

augusto2112 reviewed (approved):
LGTM if dsymutil keeps the retained types.

adrian-prantl committed:
[Debug Info] Sink type alias check into DbgTy.isForwardDecl (NFC)

adrian-prantl committed:
[Debug Info] Remove the mostly unused FragmentType from DebugTypeInfo

This cleanup allows for more consistent debug info emission since we depend less
on IRGen magic to emit debug info for types.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test
